### ISSUE TITLE
numba 0.57.0rc1 segfaults: missing ldexpf for np.exp2 on Windows

### ISSUE THREAD

**Issue by sklam** [numba/numba#8898]
from https://github.com/numba/numba/issues/8841#issuecomment-1506280834:

Original message:

I am finding that numba 0.57.0rc1 segfaults on Windows when installed with pip or conda for the following:

```python
import numpy as np
from numba import njit

f = njit(lambda x: np.exp2(x))
f(np.int8(2))
```
I create the environments via
```
conda create -n py310pip -c conda-forge python=3.10
pip install numba==0.57.0rc1
```
and
```
conda create -n py310conda -c numba -c conda-forge python=3.10 numba
```
I have not found any other function for which this occurs.

I discovered this when testing [`python-graphblas`](https://github.com/python-graphblas/python-graphblas/issues/431). Here's an example traceback:
<details>
<summary>LLVM ERROR: Symbol not found: ldexpf</summary>

```
LLVM ERROR: Symbol not found: ldexpf
Fatal Python error: Aborted

Current thread 0x00000bb0 (most recent call first):
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\llvmlite\binding\ffi.py", line 152 in __call__
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\llvmlite\binding\executionengine.py", line 92 in finalize_object
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\numba\core\codegen.py", line 1060 in wrapper
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\numba\core\codegen.py", line 999 in _finalize_specific
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\numba\core\codegen.py", line 797 in _finalize_final_module
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\numba\core\codegen.py", line 765 in finalize
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\numba\core\codegen.py", line 567 in _ensure_finalized
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\numba\core\codegen.py", line 989 in get_pointer_to_function
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\numba\core\ccallback.py", line 75 in compile
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\numba\core\compiler_lock.py", line 35 in _acquire_compile_lock
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\numba\core\decorators.py", line 311 in wrapper
  File "D:\a\python-graphblas\python-graphblas\graphblas\core\operator\unary.py", line 235 in _build
  File "D:\a\python-graphblas\python-graphblas\graphblas\core\operator\unary.py", line 305 in register_new
  File "D:\a\python-graphblas\python-graphblas\graphblas\unary\numpy.py", line 152 in __getattr__
  File "D:\a\python-graphblas\python-graphblas\graphblas\tests\test_numpyops.py", line 83 in test_npunary
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\python.py", line 194 in pytest_pyfunc_call
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\python.py", line 1799 in runtest
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\runner.py", line 169 in pytest_runtest_call
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\runner.py", line 262 in <lambda>
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\runner.py", line 341 in from_call
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\runner.py", line 261 in call_runtest_hook
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\runner.py", line 222 in call_and_report
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\runner.py", line 133 in runtestprotocol
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\runner.py", line 114 in pytest_runtest_protocol
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\main.py", line 348 in pytest_runtestloop
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\main.py", line 323 in _main
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\main.py", line 269 in wrap_session
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\main.py", line 316 in pytest_cmdline_main
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\config\__init__.py", line 166 in main
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\_pytest\config\__init__.py", line 189 in console_main
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\pytest\__main__.py", line 5 in <module>
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\coverage\execfile.py", line 199 in run
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\coverage\cmdline.py", line 830 in do_run
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\coverage\cmdline.py", line 659 in command_line
  File "C:\Miniconda3\envs\graphblas\lib\site-packages\coverage\cmdline.py", line 943 in main
  File "C:\Miniconda3\envs\graphblas\Scripts\coverage-script.py", line 10 in <module>
graphblas/tests/test_numpyops.py::test_npunary 
```

</details>

_Originally posted by @eriknw in https://github.com/numba/numba/issues/8841#issuecomment-1506280834_

**Comment by sklam** [comment]
Confirmed crash with the above reproducer. Also confirmed that it only happens for integer `int8` and `int16`. No error for bigger integers or float inputs. Unsigned integers not tested.

**Comment by sklam** [comment]
Confirmed that the bug is not conda-forge or pip specific. Also happened on with conda default channel.

**Comment by sklam** [comment]
LLVM IR for float32 input (working):

```
  %.12 = tail call float @llvm.exp2.f32(float %arg.x)
```

for int16 input (segfaulting):

```
  %0 = sext i16 %arg.x to i32
  %ldexpf = tail call float @ldexpf(float 1.000000e+00, i32 %0)
```

**Comment by sklam** [comment]
0.56 used to rely on `npy_exp2f` symbol shipped internally. Problem likely started because of the change in #8544 to use a llvm intrinsic instead; and, llvm incorrectly assumed that `ldexpf` existed and selected that lowering path.

I think this can be fixed by force promotion of integer size < 32-bit to 32-bit in that function.

**Comment by sklam** [comment]
It turns out the problem is more complicated. Numba is emitting:

```llvm
; np.exp2(int8)
  %".12" = sitofp i8 %"arg.x" to float     ; cast i8 to float
  %".13" = call float @"llvm.exp2"(float %".12")
```
 
so LLVM's optimization is responsible by recognizing the original input data type has less precision. I don't think we can simply cast to a bigger type because LLVM will easily recognizing the promotion is not necessary.

**Comment by sklam** [comment]
LLVM's `LibCallSimplifier::optimizeExp2` is responsible for this optimization:

https://github.com/llvm/llvm-project/blob/f28c006a5895fc0e329fe15fead81e37457cb1d1/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp#L1801-L1808

TLI should have set `ldexpf` as unavailable on windows: https://github.com/llvm/llvm-project/blob/f28c006a5895fc0e329fe15fead81e37457cb1d1/llvm/lib/Analysis/TargetLibraryInfo.cpp#L277

**Comment by sklam** [comment]
Given a `chk.ll` containing reproducer:

```llvm
; filename: chk.ll
; ModuleID = '<lambda>'
source_filename = "<string>"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-windows-msvc"

declare float @llvm.exp2.f32(float %casted)

define float @foo(i16 %arg) {
entry:
  %casted = sitofp i16 %arg to float
  %ret = call float @llvm.exp2.f32(float %casted)
  ret float %ret
}
```

Running `opt -instcombine` do not produce the `ldexpf`:
```bash
$ opt -instcombine chk.ll -S
```

```
; ModuleID = 'chk.ll'
source_filename = "<string>"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-windows-msvc"

; Function Attrs: nofree nosync nounwind readnone speculatable willreturn
declare float @llvm.exp2.f32(float) #0

define float @foo(i16 %arg) {
entry:
  %casted = sitofp i16 %arg to float
  %ret = call float @llvm.exp2.f32(float %casted)
  ret float %ret
}

attributes #0 = { nofree nosync nounwind readnone speculatable willreturn }
```

However, doing the same thing in llvmlite produces the `ldexpf`:

```python
# filename: chk.py
from llvmlite import binding as llvm
llvm.initialize_all_targets()

with open("chk.ll", "r") as fout:
    mod = llvm.parse_assembly(fout.read())

target = llvm.Target.from_triple(mod.triple)
tm = target.create_target_machine()

pm = llvm.ModulePassManager()
tm.add_analysis_passes(pm)

pm.add_instruction_combining_pass()
pm.run(mod)

print(mod)
```

```
; ModuleID = '<string>'
source_filename = "<string>"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-windows-msvc"

; Function Attrs: nofree nosync nounwind readnone speculatable willreturn
declare float @llvm.exp2.f32(float) #0

define float @foo(i16 %arg) {
entry:
  %0 = sext i16 %arg to i32
  %ldexpf = call float @ldexpf(float 1.000000e+00, i32 %0)
  ret float %ldexpf
}

; Function Attrs: nofree willreturn
declare float @ldexpf(float, i32 signext) #1

attributes #0 = { nofree nosync nounwind readnone speculatable willreturn }
attributes #1 = { nofree willreturn }
```

I think what we need is something like this:
https://github.com/llvm/llvm-project/blob/f28c006a5895fc0e329fe15fead81e37457cb1d1/llvm/tools/opt/opt.cpp#L732-L733

**Comment by sklam** [comment]
> I think what we need is something like this:
> https://github.com/llvm/llvm-project/blob/f28c006a5895fc0e329fe15fead81e37457cb1d1/llvm/tools/opt/opt.cpp#L732-L733

Confirmed with https://github.com/numba/llvmlite/pull/934
and

```python
from llvmlite import binding as llvm
llvm.initialize()
llvm.initialize_all_targets()
llvm.initialize_all_asmprinters()

with open("chk.ll", "r") as fout:
    mod = llvm.parse_assembly(fout.read())

target = llvm.Target.from_triple(mod.triple)
tm = target.create_target_machine()

pm = llvm.ModulePassManager()
tm.add_analysis_passes(pm)
pm.add_target_library_info(mod.triple)  # <---- NEW


pm.add_instruction_combining_pass()
pm.run(mod)

print(mod)
```

**Comment by ZupoLlask** [comment]
Thanks for your hard work sorting this issue, @sklam! 🙂

**Comment by stuartarchibald** [comment]
Closing, fixed by #8907.
